daffodil-io/src/main/scala/org/apache/daffodil/io/Dump.scala

```diff
@@ -516,24 +544,41 @@ class DataDumper {
           // Either way, we got our one character
           Assert.invariant(nConsumedBytes > 0)
           Assert.invariant(cb.hasArray)
-          var allChars = cb.array
-          remapped = if (allChars.length > 1) allChars.mkString else Misc.remapCodepointToVisibleGlyph(allChars(0)).toChar.toString
-          nCols = if (allChars.length > 1) {
-            try {
-              var uCodePoint = UCharacter.getCodePoint(allChars(0), allChars(1))
-              charNColumns(uCodePoint)
-            } catch {
-              case e: IllegalArgumentException => {
-                allChars.mkString.length
+          val allChars = cb.array
+
+          val uCodePoint =
+            if (allChars.length > 1) {
+              try {
+                UCharacter.getCodePoint(allChars(0), allChars(1))
+              } catch {
+                case e: IllegalArgumentException => {
+                  -1
+                }
+              }
+            } else allChars(0)
+
+          val (r: String, n: Int) =
+            if (allChars.length > 1) {
+              if (uCodePoint == -1) {
+                allChars.map(c => homogenizeChars(c)).foldLeft(("", 0)) {
```

Comment:
I don't understand this piece of code. Please add comment as to what this case is doing.

Reply:
I updated the code for clarity, adding comments and doing some additional checks so we don't need to rely on exception handling for illegal surrogate pairs!